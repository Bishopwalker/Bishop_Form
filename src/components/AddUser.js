import React, {useState, useEffect} from 'react';
import { useDispatch, useSelector} from 'react-redux';
import {addPerson} from'../ducks/reducer'
 import {dispatchInfo,addPersonFunction, initializedFunction} from "../ducks/actions";
import '../App.css';
import {Link,useNavigate,useLocation} from "react-router-dom";
import axios from "axios";


const AddUser=(props)=>{
    const statePerson = useSelector(state=>state.person.id);
    const [submit,setSubmit] = useState(false)
    const dispatch = useDispatch();
    const hobbyRef=React.useRef(),
        firstnameRef = React.useRef();

    const [firstname,setFirstName] = useState(''),
        [lastname,setLastName] = useState(''),
        [person_age,setAge] = useState(''),
        [hobbies,setHobbies] = useState(''),
        [id,setID] = useState(),
        [person,setPerson] = useState({
            firstname,
            lastname,
            person_age,
            hobbies
        }),
        [isMounted,setIsMounted] = useState(false),
        [errors,setErrors] = useState([]),
        [showErrorMessage,setShowErrorMessage] = useState(false),
        [classChoice,setClassChoice] = useState('reset1')

    const error={
       name:'Your first or last name need to be less than 50 characters',
        name_required:'You must enter a first and last name',
        person_age:'Your Age is must be between 1-110 years Old',
        person_age_min:'Your Min person_age must be 18',
        person_age_required:'Please enter a valid person_age over 18',
        hobby_min:'Your Hobbies must be at least 5 characters long',
        hobby_max:'Your Hobbies cannot be longer than 250 characters',
        hobby_required: 'Please include a hobby'
    }
    const fixStr=(str)=>{
            let firstLetter = str.substring(0,1).toUpperCase() + str.substring(1).toLowerCase()

            if(!str){
            setErrors((prev)=>[...prev,error.name_required])
        }

       else if(firstLetter.length > 50) {

          setErrors((prev)=>[...prev,error.name])

        }

        return firstLetter.replace(/\s/g,"_")
    },
        fixAge=(person_age)=>{
        if(person_age>110) {
            setErrors((prev)=>[...prev,error.person_age])
        }
            if(!person_age){
                setErrors((prev)=>[...prev,error.person_age_required])
            }
       else if(person_age<18 && person_age !==undefined){
            setErrors((prev)=>[...prev,error.person_age_min])
        }

        return person_age
        },
        fixHobbies=(hobby)=>{
        if(hobby.length > 250){
            setErrors((prev)=>[...prev,error.hobby_max])
        }
        if(!hobby){
            setErrors((prev)=>[...prev,error.hobby_required])
        }
        else if(hobby.length<5){
            setErrors((prev)=>[...prev,error.hobby_min])
        }
        return hobby
        };

// console.log(errors.length)
    const addNewPerson = async() => {

                  await setID(statePerson)
                  await setPerson({
                      id: id,
                      firstname: fixStr(firstname),
                      lastname: fixStr(lastname),
                      person_age: fixAge(person_age),
                      hobbies: fixHobbies(hobbies)
                  })
    }
const updatePerson=async()=>{

        const results = await axios.put(`http://localhost:3003/auth/register/${id}`, person)
    console.log(results);

}
const isLoggedIn=async()=>{
        const users = await axios.get('http://localhost:3003/auth/check-user')

}
useEffect(()=>{
    firstnameRef.current.focus()

},[])
    useEffect(async()=>{
       dispatch(initializedFunction())

if(isMounted) {
if(errors.length > 0){

  await  setShowErrorMessage(true)
    firstnameRef.current.focus()
    setClassChoice('reset2')


}else {
    setSubmit(true)
    //dispatch(addPersonFunction(person))

    setErrors([])
    setShowErrorMessage(false)
    setTimeout(() => {
        setSubmit(false)
    }, 750)
    return async()=> {
        await setIsMounted(false)
        setClassChoice('reset1')

    }
}
}
    },[person])

    const reset=(e)=>{
        e.preventDefault()
        if(firstname.length ===0 && lastname.length === 0 && person_age.length === 0 && hobbies.length === 0){
            window.alert(`Cannot clear an empty field`)
        }else{
            setFirstName('')
            setLastName('')
            setAge('')
            setHobbies('')

        }
        firstnameRef.current.focus()
    }


    const handleFirstNameChange=async(e)=>{
        e.preventDefault();
     // const {firstname, lastname, person_age, hobbies} = e.target;
        setErrors([]);
        setShowErrorMessage(false)
      await  setFirstName(e.target.value)

    },
        handleLastNameChange=(e)=>{
        e.preventDefault();
            setErrors([]);
            setShowErrorMessage(false)
        setLastName(e.target.value)
        },
        handleAgeChange=(e)=>{
        e.preventDefault();
            setErrors([]);
            setShowErrorMessage(false)
        setAge(e.target.value)
    },
        handleHobbiesChange=(e)=>{
        e.preventDefault();
            setErrors([]);
            setShowErrorMessage(false)
        setHobbies(e.target.value)
        }


    const handleSubmit = async(e) => {
        e.preventDefault();
if(errors.length===0) {
    await setIsMounted(true)
    await addNewPerson()

    setFirstName('')
    setLastName('')
    setAge('')
    setHobbies('')
}else{
    firstnameRef.current.focus()
}


    }

    return(
        <div className="App-header outSide" >
            <div id='contain'>
                <div id='right'>
            {showErrorMessage ?(
                <div>
                    <h3>Errors</h3>
                    {errors.map((error, index) => (
                        <li key={index}>{error}</li>
                    ))}
                </div>
            ):null}
                </div>
        <form onSubmit={handleSubmit}>
                    <span id='spans'><h2>Fill Out Form</h2></span>

    <div id='form' >
            <input
                type='text'
                name={firstname}
                placeholder="First Name"
                onChange={handleFirstNameChange}
                value={firstname}
                ref={firstnameRef}

            />
            <input
                type='text'
                name={lastname}
                placeholder="Last Name"
                onChange={handleLastNameChange}
                value={lastname}
            />
            <input
                type={'number'}
                name={person_age}
                placeholder="Age"
                onChange={handleAgeChange}
                value={person_age}
            />
            <input
                type={'text'}
                name={hobbies}
                placeholder="Hobbies"
                onChange={handleHobbiesChange}
                value={hobbies}
                ref={hobbyRef}
            />
        <Link to='/viewDatabase'>
            <button>View Database</button>
        </Link>
    </div>

            <div>
            <button className={classChoice} onClick={reset}>Reset</button>
            <button >Submit Form</button>
            </div>
        </form>
            <Link to='/viewContainer'  >
            <button id='listButton'>See Local List</button>
            </Link>
            {submit?(<div className='green'><h2>You Have Successfully added a Person!</h2></div>):null}
        </div>
        </div>
    )
}


export default AddUser