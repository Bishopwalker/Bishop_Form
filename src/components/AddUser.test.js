import React from 'react';
import{render, fireEvent, waitFor, screen} from '@testing-library/react';
import AddUser from './AddUser';
import * as reactRedux  from 'react-redux';
import {createMemoryHistory} from "history";
import {Router} from "react-router-dom";
import '@testing-library/jest-dom';
import userEvent from "@testing-library/user-event";
import {BrowserRouter} from "react-router-dom";

describe("Add User",()=>{
    const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');
    const useDispatchMock = jest.spyOn(reactRedux, 'useDispatch');

    beforeEach(()=>{
        useSelectorMock.mockClear();
        useDispatchMock.mockClear();


    })
    //
    // it('Should take a snapshot',async()=>{
    //     try {
    //         const {asFragment} = await render(<AddUser/>, {wrapper: BrowserRouter});
    //         expect(asFragment()).toMatchSnapshot();
    //     } catch (e) {
    //         console.log(e);
    //     }
    //
    // })

it('does something',()=> {
    const dummyDispatch = jest.fn();
    const dummySelector = jest.fn();
    useDispatchMock.mockReturnValue(dummyDispatch);
    useSelectorMock.mockReturnValue(dummySelector);
    //Sanity Check

    expect(dummyDispatch).not.toHaveBeenCalled();
    render(<AddUser/>, {wrapper: BrowserRouter});
    //screen.getByRole('button',{name:'Submit Form'});
    userEvent.click(screen.getByRole('button', {name: 'Submit Form'}));
    expect(dummyDispatch).toHaveBeenCalled();   //Check if the dispatch function is called
})

})