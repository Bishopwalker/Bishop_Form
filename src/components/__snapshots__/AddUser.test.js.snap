// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Add User Should take a snapshot 1`] = `
<DocumentFragment>
  <div
    class="App-header outSide"
  >
    <div
      id="contain"
    >
      <div
        id="right"
      />
      <form>
        <span
          id="spans"
        >
          <h2>
            Fill Out Form
          </h2>
        </span>
        <div
          id="form"
        >
          <input
            name=""
            placeholder="First Name"
            type="text"
            value=""
          />
          <input
            name=""
            placeholder="last Name"
            type="text"
            value=""
          />
          <input
            name=""
            placeholder="age"
            type="number"
            value=""
          />
          <input
            name=""
            placeholder="Hobbies"
            type="text"
            value=""
          />
        </div>
        <div>
          <button
            class="reset1"
          >
            Reset
          </button>
          <button>
            Submit Form
          </button>
        </div>
      </form>
      <a
        href="/viewContainer"
      >
        <button
          id="listButton"
        >
          See List
        </button>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
